### PR TITLE
Fixing a bug with treeselector and simplifying filter usage

### PR DESCRIPTION
**Pull request by vadym-hyryn**
[Bug 35594](https://dev.azure.com/orckestra001/OrckestraCommerce/_workitems/edit/35594/)
Method GetPageIdFromDataToken of GetPageId class was parsing page ID from entity token. But depending on different situations (add datafolder element, edit datafolder element, add metadata data) there will be different entity tokens. I've implemented all these 3 cases, so now filtering for adding/editing datafolders will work as well. If there are another cases except of datafolders and metadata fields, they should be also implemented.

Initially in the widget was an option to add page selector and to configure different filtering options: Level 1 (home page), level 2, level 3, level 4. But as we have to display a tree element to the end user, and a tree element should have some root, and on level 2 there could be 2,3,4,5 subtrees, there was no way to use these options level 2,3 and 4 and they didn't work as well. So there left only 1 available and possible for use option: level 1 - to show web pages only of some certain webpage. To leave a combobox with only 1 option "level 1" seemed not very good, therefore to PageReferenceSelectorWidgetFunctionBase instead of configuring option with levels was added a bool option: to filter or not to filter pages by active website. Because a standard bool widget provider was used, class HomePageSelectorWidgetFunction is not necessary anymore so it was deleted.

### CODE REVIEW

**Comment by napernik** [comment]
**Composite.Pages.GetPageId** function existed for a long time and is used not only in the newly added tree selector, but in some of the XSLT functions of our packages (mostly for reading current page id), as well as numerous customer solutions. Renaming it or removing some of its features would break compatibility with lots of existing code.

**Comment by vadym-hyryn** [comment]
> **Composite.Pages.GetPageId** function existed for a long time and is used not only in the newly added tree selector, but in some of the XSLT functions of our packages (mostly for reading current page id), as well as numerous customer solutions. Renaming it or removing some of its features would break compatibility with lots of existing code.

Thank you a lot, I had to notice, updated with current GetPageId usage

File: Composite/Plugins/Functions/WidgetFunctionProviders/StandardWidgetFunctionProvider/DataReference/PageReferenceSelectorWidgetFunctionBase.cs [hunk 1]
```diff
@@ -1,27 +1,36 @@
...
+                "Enabled: filter pages by active website",
+                "Disabled: no filter, show all websites pages (default)");
+
+            base.AddParameterProfile(new ParameterProfile(
+                HomePageIdParameterName,
+                typeof(bool),
+                false,
+                new ConstantValueProvider(false),
+                boolProvider,
+                "Filter by website",
```
Comment (neexite) [inline Composite/Plugins/Functions/WidgetFunctionProviders/StandardWidgetFunctionProvider/DataReference/PageReferenceSelectorWidgetFunctionBase.cs]: Why you renname "Filter by Home Page" to "Filter by website"

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs [hunk 2]
```diff
@@ -70,30 +71,45 @@ private Guid GetCurrentPageId()
...
             {
-                return GetPageIdFromDataToken(entityToken as DataEntityToken);
+                return GetPageIdFromDataToken(entityToken);
             }
 
             return null;
         }
 
-        private Guid? GetPageIdFromDataToken(DataEntityToken entityToken)
+        private Guid? GetPageIdFromDataToken(object entityTokenObj)
```
Comment (neexite) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Can we leave EntityToken instead object?

Comment (neexite) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: If we get pageId from EntityToken we should rename to GetPageIdFromEnitityToken

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs [hunk 3]
```diff
@@ -70,30 +71,45 @@ private Guid GetCurrentPageId()
...
 
-            var data = entityToken.Data as IPage;
-            var pageId = data?.Id;
+            if (entityTokenObj is DataEntityToken dataEntityToken)
+            {
+                if (dataEntityToken.Data is IPage cPage)
+                {
+                    pageId = cPage?.Id;
+                }
+                else if (dataEntityToken.Data is IPageRelatedData cPagerelatedDate)
```
Comment (neexite) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: *page**R**elatedDat**a**

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs [hunk 4]
```diff
@@ -70,30 +71,45 @@ private Guid GetCurrentPageId()
...
 
-            if (Type.GetType(entityToken.Type, throwOnError: false) != typeof(IPage))
-                return null;
+            Guid? pageId = null;
 
-            var data = entityToken.Data as IPage;
-            var pageId = data?.Id;
+            if (entityTokenObj is DataEntityToken dataEntityToken)
+            {
+                if (dataEntityToken.Data is IPage cPage)
```
Comment (neexite) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: *page

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs [hunk 5]
```diff
@@ -70,30 +71,45 @@ private Guid GetCurrentPageId()
...
+                    pageId = cPage?.Id;
+                }
+                else if (dataEntityToken.Data is IPageRelatedData cPagerelatedDate)
+                {
+                    pageId = cPagerelatedDate?.PageId;
+                }
+            }
+            else if (typeof(IPage).IsAssignableFrom(Type.GetType(entityToken.Type, throwOnError: false)))
+            {
+                Guid.TryParse(entityToken?.Id, out Guid parcedId);
```
Comment (neexite) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Guid.TryParse(entityToken?.Id, out pageId); more simple

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs [hunk 6]
```diff
@@ -70,30 +71,45 @@ private Guid GetCurrentPageId()
...
+                if (dataEntityToken.Data is IPage cPage)
+                {
+                    pageId = cPage?.Id;
+                }
+                else if (dataEntityToken.Data is IPageRelatedData cPagerelatedDate)
+                {
+                    pageId = cPagerelatedDate?.PageId;
+                }
+            }
+            else if (typeof(IPage).IsAssignableFrom(Type.GetType(entityToken.Type, throwOnError: false)))
```
Comment (neexite) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Can you describe the sample for this?

File: Composite/Plugins/Functions/WidgetFunctionProviders/StandardWidgetFunctionProvider/DataReference/PageReferenceSelectorWidgetFunctionBase.cs [hunk 7]
```diff
@@ -43,15 +52,22 @@ public override XElement GetWidgetMarkup(ParameterList parameters, string label,
...
             var element = new XElement(selector.Name.Namespace + "DataReferenceTreeSelector.SearchToken",
                 new XElement(f + "StaticMethodCall",
                     new XAttribute("Type", TypeManager.SerializeType(typeof(PageReferenceSelectorWidgetFunctionBase))),
                     new XAttribute("Method", nameof(GetPageSearchToken)),
-                    new XElement(f + "StaticMethodCall.Parameters", parameter)));
+                    new XElement(f + "StaticMethodCall.Parameters",
+                        new XElement(ft + "function",
+                            new XAttribute("name", "Composite.Pages.GetPageId"),
+                            new XElement(ft + "param",
+                                new XAttribute("name", "SitemapScope"),
```
Comment (neexite) [inline Composite/Plugins/Functions/WidgetFunctionProviders/StandardWidgetFunctionProvider/DataReference/PageReferenceSelectorWidgetFunctionBase.cs]: nameof(SitemapScope)

File: Composite/Plugins/Functions/WidgetFunctionProviders/StandardWidgetFunctionProvider/DataReference/PageReferenceSelectorWidgetFunctionBase.cs (hunk 1 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/WidgetFunctionProviders/StandardWidgetFunctionProvider/DataReference/PageReferenceSelectorWidgetFunctionBase.cs]: It seemed for me more clear for understanding. Filter by website => so we can expect to see pages of some certain website. Filter by homepage => so we can expect to see pages which a) have in a **root** a home page, so we will see pages of a certain website or b) have in a **parent**  a home page, so we will see level 1 and level 2 pages. It seemed for me a little bit misunderstanding, but it is just my suggestion. Keep it or turn back?

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs (hunk 2 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Updated

Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Right, thanks:) Renamed

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs (hunk 3 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Renamed, thanks

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs (hunk 4 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Renamed

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs (hunk 5 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Yes, it  would be good, but pageId is NullAble there

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs (hunk 6 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Yes, sure

- `if (dataEntityToken.Data is IPage page)` - appears while adding a metadata element;
- `else if (dataEntityToken.Data is IPageRelatedData pageRelatedData)` - appears while **editing** a datafolder element;
- `else if (typeof(IPage).IsAssignableFrom(Type.GetType(entityToken.Type, throwOnError: false)))` - appears while **adding** a datafolder element.

I added explaining comments to original code

File: Composite/Plugins/Functions/WidgetFunctionProviders/StandardWidgetFunctionProvider/DataReference/PageReferenceSelectorWidgetFunctionBase.cs (hunk 7 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/WidgetFunctionProviders/StandardWidgetFunctionProvider/DataReference/PageReferenceSelectorWidgetFunctionBase.cs]: Thanks, updated, also updated in GetPageIdFunction class

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs [hunk 8]
```diff
@@ -71,40 +71,43 @@ private Guid GetCurrentPageId()
...
 
-            if (entityTokenObj is DataEntityToken dataEntityToken)
+            if (entityToken is DataEntityToken dataEntityToken)
             {
-                if (dataEntityToken.Data is IPage cPage)
+                //appears while adding a metadata element
+                if (dataEntityToken.Data is IPage page)
                 {
-                    pageId = cPage?.Id;
+                    pageId = page?.Id;
```
Comment (burningice2866) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: `page` cannot be null here so you can omit the ?

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs [hunk 9]
```diff
@@ -71,40 +71,43 @@ private Guid GetCurrentPageId()
...
                 }
-                else if (dataEntityToken.Data is IPageRelatedData cPagerelatedDate)
+                //appears while editing a datafolder element
+                else if (dataEntityToken.Data is IPageRelatedData pageRelatedData)
                 {
-                    pageId = cPagerelatedDate?.PageId;
+                    pageId = pageRelatedData?.PageId;
```
Comment (burningice2866) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: pageRelatedData cannot be null here so you can omit the ?

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs [hunk 10]
```diff
@@ -24,7 +24,7 @@ public GetPageIdFunction(EntityTokenFactory entityTokenFactory)
 
         public override object Execute(ParameterList parameters, FunctionContextContainer context)
         {
-            if (parameters.TryGetParameter<SitemapScope>("SitemapScope", out var sitemapScope) == false)
+            if (parameters.TryGetParameter<SitemapScope>(nameof(SitemapScope), out var sitemapScope) == false)
```
Comment (burningice2866) [inline Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: `if (!parameters.TryGetParameter<SitemapScope>(nameof(SitemapScope), out var sitemapScope))`

?

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs (hunk 8 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Thanks, omitted

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs (hunk 9 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: Omitted 👍

File: Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs (hunk 10 above)
Comment (vadym-hyryn) [reply Composite/Plugins/Functions/FunctionProviders/StandardFunctionProvider/Pages/GetPageIdFunction.cs]: I didn't notice this when was updating this line, thanks, changed

**Pull request merged**